### PR TITLE
Fix PR #15489 (Dynamic Library Loading Support)

### PR DESCRIPTION
**Pull request by mseth10**
## Description ##
This PR builds on top of https://github.com/apache/incubator-mxnet/pull/15489, which was reverted by https://github.com/apache/incubator-mxnet/pull/15755.

It includes the following additional changes:
- Changes to Makefile and CMake flows to build sample library as part of MXNet build
- Modifications to CI to stash/un-stash the library binary from build stage to test stage
- Loads the binary as part of unit test

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/unittest/test_library_loading.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def test_library_loading():
     elif (os.name=='nt'):
         lib = 'mylib.dll'
 
-    fname = mx.test_utils.download('https://mxnet-demo-models.s3.amazonaws.com/lib_binary/'+lib)
-    fname = os.path.abspath(fname)
+    # fname = mx.test_utils.download('https://mxnet-demo-models.s3.amazonaws.com/lib_binary/'+lib)
+    # fname = os.path.abspath(fname)
+    fname = 'Somewhere/In/The/CI'
```
Comment (mseth10) [inline tests/python/unittest/test_library_loading.py]: @marcoabreu what should we change this to to get the path to the library libsample_lib.so/.dll?

Comment (marcoabreu) [reply tests/python/unittest/test_library_loading.py]: Check the environment variables. I think there's one called "WORKSPACE" which points to the folder that is allocated for that job

File: example/lib_api/Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+all:
+	g++ -shared -fPIC mylib.cc -o mylib.so -I ../../include/mxnet
```
Comment (larroy) [inline example/lib_api/Makefile]: If CI doesn't pass, I suggest to use $(CC) and $(CXX), if not it's not worth another CI round.

Comment (samskalicky) [reply example/lib_api/Makefile]: @larroy this isnt building in the CI. this is just an example

File: src/common/library.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+/*!
+ * \brief a templated function that fetches from the library
+ * a function pointer of any given datatype and name
+ * \param T a template parameter for data type of function pointer
+ * \param lib library handle
+ * \param func_name function name to search for in the library
+ * \return func a function pointer
+ */
+template<typename T>
+T get_func(void *lib, char *func_name) {
```
Comment (larroy) [inline src/common/library.h]: We don't want to get funky inside a mxnet namespace?

File: src/engine/threaded_engine.h [hunk 4]
```diff
@@ -331,6 +332,10 @@ class ThreadedEngine : public Engine {
       kill_.store(true);
     }
     finished_cv_.notify_all();
+    // close opened libraries
+    for (auto const& lib : loaded_libs) {
+      close_lib(lib.second);
```
Comment (larroy) [inline src/engine/threaded_engine.h]: why this in the engine? I suggest to do global cleanup in initialize.cc.

related: https://github.com/apache/incubator-mxnet/pull/15762

File: example/lib_api/Makefile (hunk 2 above)
Comment (marcoabreu) [reply example/lib_api/Makefile]: Well but it should be - how else it going to be tested?

Comment (larroy) [reply example/lib_api/Makefile]: I understand, but still this is a PITA if a Makefile doesn't respect the compiler from the environment. That's why I said is not worth another CI round. If you make other changes I would change it. Or is there a reason not to?

Comment (samskalicky) [reply example/lib_api/Makefile]: How does that work for windows? will it automatically have CXX be "cl" ?

File: src/engine/threaded_engine.h (hunk 4 above)
Comment (mseth10) [reply src/engine/threaded_engine.h]: addressed.

Comment (mseth10) [reply src/engine/threaded_engine.h]: addressed

File: Makefile [hunk 5]
```diff
@@ -576,7 +576,7 @@ lib/libmxnet.a: $(ALLX_DEP)
 	@mkdir -p $(@D)
 	ar crv $@ $(filter %.o, $?)
 
-lib/libmxnet.so: $(ALLX_DEP)
+lib/libmxnet.so: $(ALLX_DEP) sample_lib
```
Comment (zachgk) [inline Makefile]: If it is not actually a dependency of libmxnet.so, we shouldn't put it here. Can we move it to .PHONY and/or all target instead?

Comment (mseth10) [reply Makefile]: Addressed. Moved to all target.

File: src/common/library.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+#ifndef MXNET_COMMON_LIBRARY_H_
+#define MXNET_COMMON_LIBRARY_H_
+
+#include <iostream>
+#include <map>
+#include <string>
+#include "dmlc/io.h"
+
+// map of libraries loaded
+static std::map<std::string, void*> loaded_libs;
```
Comment (larroy) [inline src/common/library.h]: This should not be static in a header. I will change it in my PR. This creates a symbol in every compilation unit.

File: Makefile [hunk 7]
```diff
@@ -659,6 +659,9 @@ cpplint:
 pylint:
 	python3 -m pylint --rcfile=$(ROOTDIR)/ci/other/pylintrc --ignore-patterns=".*\.so$$,.*\.dll$$,.*\.dylib$$" python/mxnet tools/caffe_converter/*.py
 
+sample_lib:
+	$(CXX) -shared -fPIC example/lib_api/mylib.cc -o libsample_lib.so -I include/mxnet
```
Comment (TaoLv) [inline Makefile]: `libsample_lib.so` cannot be cleaned via `make clean`. Could you please move it to the lib folder?

Comment (samskalicky) [reply Makefile]: I dont think we want to move it to the lib folder, since this library is not used for linking or typical LD_LIBRARY_PATH stuff. What if we just add a rm line to the clean target to delete this? We can do it as part of the custom op PR #15921

Comment (TaoLv) [reply Makefile]: Thank you for the explanation, @samskalicky . I'm just wondering if we should put a `.so` file in the root folder and I feel difficult to understand the usage of the `.so` from its name.

Comment (samskalicky) [reply Makefile]: @TaoLv this has been fixed in the custom Op PR from @rondogency by adding entries to the clean target:
https://github.com/apache/incubator-mxnet/pull/15921/files#diff-b67911656ef5d18c4ae36cb6741b7965R763

Comment (TaoLv) [reply Makefile]: Thank you for fixing that.
